### ISSUE TITLE
Subscription intermittently stops delivering messages

### ISSUE THREAD

**Issue by mikeb01** [real-logic/Aeron#244]
Hi, similar issue to #234, but with slightly different behaviour, so I'm raising a new issue.  One of the subscriptions on a specific channel causes all of the others to stall and results in back pressure on the publication.

Here is AeronStat output for stream `-1762877036`.  Note registrationId `1167` is show a sub-pos much smaller than the others.

AeronStat
```
983:               65,184 - pub-lmt: 853 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
984:               32,416 - snd-pos: 853 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1099:               29,920 - sub-pos: 1428 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1134:               29,920 - sub-pos: 460 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1135:               29,920 - sub-pos: 428 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1136:               29,920 - sub-pos: 309 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1137:               29,920 - sub-pos: 241 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1140:               29,920 - sub-pos: 197 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1141:               29,920 - sub-pos: 76 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1227:               29,920 - sub-pos: 1332 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1435:               29,920 - sub-pos: 718 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1436:               29,920 - sub-pos: 776 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1437:                  384 - sub-pos: 1167 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1438:               32,416 - rcv-hwm: 1270 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1464:               29,920 - sub-pos: 1293 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1712:               29,920 - sub-pos: 1962 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1761:               29,920 - sub-pos: 1671 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
1973:               29,920 - sub-pos: 1836 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
2039:               29,920 - sub-pos: 1928 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
2060:               29,920 - sub-pos: 2038 880428294 -1762877036 aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
```

I have some custom monitoring of my handler so that I can track if this subscription is still being polled.  You can see the streamId and registrationId matches the problematic subscription and pollCount is increasing, which means I'm confident our polling code hasn't crashed.

```
$>bean com.lmax:name=tfx-group__InstrumentNotificationServicePublic,service=cfdx,type=AeronReceiver 
#bean is set to com.lmax:name=tfx-group__InstrumentNotificationServicePublic,service=cfdx,type=AeronReceiver
$>get *
#mbean = com.lmax:name=tfx-group__InstrumentNotificationServicePublic,service=cfdx,type=AeronReceiver:
RegistrationId = 1167;
TopicName = tfx-group::InstrumentNotificationServicePublic;
StreamId = -1762877036;
PollCount = 441575;
Uri = aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1;

$>get *
#mbean = com.lmax:name=tfx-group__InstrumentNotificationServicePublic,service=cfdx,type=AeronReceiver:
RegistrationId = 1167;
TopicName = tfx-group::InstrumentNotificationServicePublic;
StreamId = -1762877036;
PollCount = 691272;
Uri = aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1;
```

Polling code:

        public int poll()
        {
            final int poll = subscription.poll(handler, 20);
            ++pollCount;
            return poll;
        }

Looking at the log inspect for the appropriate subscription image show that aeron thinks the term is unused and dirty.  I.e. the first show no messages, but has not been cleaned the remaining 2 terms are all zeros.  The LogInspector suggests that there are no messages in that term.

```
 ======================================================================
 Fri Jun 17 03:38:17 UTC 2016 Inspection dump for /dev/shm/aeron-tradex/images/UDP-ac1d0101-0-efc10001-10000-347A4506-96ECA194-4F6.logbuffer
 ======================================================================
 Time of last SM: Thu Jan 01 00:00:00 UTC 1970
 Initial term id: -1970406852
    Active index: 0
     Term length: 65536
      MTU length: 4096

 default Data Header{frame_length=0 version=0 flags=11000000 type=1 frame_length=0 term_offset=0 session_id=880428294 stream_id=-1762877036 term_id=-1970406852 reserved_value=0}

 Index 0 Term Meta Data status=CLEAN termOffset=0 termId=0
 Index 1 Term Meta Data status=CLEAN termOffset=0 termId=0
 Index 2 Term Meta Data status=CLEAN termOffset=0 termId=0
 %n======================================================================
 Index 0 Term Data

 Data Header{frame_length=0 version=0 flags=00000000 type=0 frame_length=0 term_offset=0 session_id=0 stream_id=0 term_id=0 reserved_value=0}
 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
 00002C00000000C001008001000006457A3494A1EC963CFA8D8A0000000000000000000000080000000000000000000000000000000000000000000000000000000020000000
 00C00100C001000006457A3494A1EC963CFA8D8A00000000000000002C00000000C00100E001000006457A3494A1EC963CFA8D8A000000000000000000000008000000000000
 ...
```

**Comment by mjpt777** [comment]
It looks like the log is in its initial term and has been half filled by the position counters. However the termOffset suggests the tail has not been incremented which is just wrong. I'd like a closer look at that log buffer. Can you send it?

**Comment by mjpt777** [comment]
@mikeb01 Are the client subscribers in the same process or different processes?

**Comment by tmontgomery** [comment]
The min of the sub-pos will be the position that is set at 384. That will be the consumption position in the SMs and from which the window will begin.

With the other sub-pos values and if that Subscription is being polled, the only way I can see that position not incrementing is if there is loss that isn't being repaired right at the point and the other subscriptions started right after the loss point.

The internal contiguous position, rebuildPosition, isn't tracked by a counter. This makes me think it might be good to make it a counter along with the rcv-hwm.

@mjpt777 Images are a little different than publication buffers. The tail isn't really used. Instead, the term reading is totally controlled by the presence of unfilled framelength values (or reaching capacity of the term while reading). The nature of Images might make the log inspector output seem strange. In any case, a closer look at the Image logbuffer might give us an idea of what has happened.

**Comment by mjpt777** [comment]
@tmontgomery You are right. I forgot that Images do not have the tail updated. They don't need it.

+1 I think it would be useful to have a rcv-pos counter that is the rebuild position for monitoring.

**Comment by tmontgomery** [comment]
I'll add the rcv-pos counter. It would be quite useful.

Nice thing is with the image logbuffer and log inspector and the stats, we could look at what is likely to have occurred. Or at least what might have occurred.

**Comment by mikeb01** [comment]
I've attached the publication and the image data ([aeron-data.zip](https://github.com/real-logic/Aeron/files/321514/aeron-data.zip)) for the specified stream.  The subscriptions and publication are in separate processes.  The media driver is also in separate process too.

**Comment by mikeb01** [comment]
Is there any other information that would be useful for debugging this issue.  I can do some re-testing on Friday.

**Comment by mjpt777** [comment]
I think it would be useful to record the joining position in the log buffer for images for later tracking.

**Comment by mjpt777** [comment]
@mikeb01 Do you have a screen grab of the system counters from this time? I see you have some from the positions.

**Comment by tmontgomery** [comment]
Let me add some more stats for tracking so that when run again, there is more information. Will look at that later today.

**Comment by mikeb01** [comment]
Having the Aeron data files sit around after shut down is very useful...

```
20:33:27 - Aeron Stat
=========================
  0:           88,291,136 - Bytes Sent
  1:          899,057,984 - Bytes received
  2:                    0 - Failed offers to ReceiverProxy
  3:                    0 - Failed offers to SenderProxy
  4:                    0 - Failed offers to DriverConductorProxy
  5:              266,102 - NAKs sent
  6:              260,676 - NAKs received
  7:           12,875,602 - Status Messages sent
  8:           12,805,048 - Status Messages received
  9:           25,018,505 - Heartbeats sent
 10:           24,203,284 - Heartbeats received
 11:              100,482 - Retransmits sent
 12:                4,978 - Flow control under runs
 13:               55,875 - Flow control over runs
 14:                    0 - Invalid packets
 15:                    0 - Errors
 16:                    0 - Short sends
 17:              493,189 - Client keep-alives
 18:                  439 - Sender flow control limits applied
 19:                    0 - Unblocked Publications
 20:                    0 - Unblocked Control Commands
 21:                    0 - Possible TTL Asymmetry
 22:                    1 - send-channel: aeron:udp?group=239.193.0.1:10000|interface=172.29.1.1
```

**Comment by mjpt777** [comment]
Thanks Mike. As an observation that machine is seeing a lot of loss. It might be worth increasing the size of the socket buffers to see if it comes more under control by compensating for the congestion once this gets resolved.

**Comment by mjpt777** [comment]
@tmontgomery Does that ratio of NAKs to retransmits look fishy to you?

**Comment by tmontgomery** [comment]
Possibly. It definitely means that less than 40% of NAKs are being answered. It could be because the retransmit is not found when scanning, though.

**Comment by mjpt777** [comment]
@mikeb01 The beginning of the term being zeros is very odd. If it happens again can you grab another image of the publication or the source publication to see if they are the same?

**Comment by mikeb01** [comment]
I plan to do some more testing tomorrow, so will do.  I can replicate this fairly easily.  Around half of the ~60 application instances in CI run into this problem.

**Comment by mjpt777** [comment]
@mikeb01 Can you post the config options you have applied?

**Comment by mikeb01** [comment]
Our setup is below.  The OS buffer memory max size is probably less than is specified.  As an aside, I wonder if it would be worthwhile having the configuration options stored as counters so they could be retrieved via AeronStat.

```
        System.setProperty("aeron.event.log", "CMD_IN_ADD_PUBLICATION,CMD_IN_ADD_SUBSCRIPTION");
        System.setProperty("aeron.socket.so_sndbuf", "2097152");
        System.setProperty("aeron.socket.so_rcvbuf", "2097152");
        System.setProperty("aeron.rcv.buffer.length", "16384");
        System.setProperty("agrona.disable.bounds.checks", "true");
        System.setProperty("aeron.term.buffer.sparse.file", "true");

        final int termBufferLength = 64 * 1024;

        final MediaDriver.Context ctx = new MediaDriver.Context()
                .threadingMode(ThreadingMode.SHARED)
                .sharedIdleStrategy(SlowIdleStrategy.INSTANCE)
                .imageLivenessTimeoutNs(10_000_000_000L)
                .publicationTermBufferLength(termBufferLength)
                .initialWindowLength(termBufferLength / 2);
```

**Comment by mjpt777** [comment]
I think it makes sense to have the properties stored in the C'n'C file as a section that can be read with a tool. They are not really counters.

**Comment by mikeb01** [comment]
I've run the 0.9.9-SNAPSHOT version through our CI system and I get the same sort of failure.

AeronStat:
```
02:35:40 - Aeron Stat
=========================
  0:           94,747,936 - Bytes sent
  1:           97,622,944 - Bytes received
  2:                    0 - Failed offers to ReceiverProxy
  3:                    0 - Failed offers to SenderProxy
  4:                    0 - Failed offers to DriverConductorProxy
  5:               37,100 - NAKs sent
  6:               37,100 - NAKs received
  7:            1,393,407 - Status Messages sent
  8:            1,393,634 - Status Messages received
  9:            2,696,965 - Heartbeats sent
 10:            2,472,784 - Heartbeats received
 11:               11,251 - Retransmits sent
 12:                1,158 - Flow control under runs
 13:                    0 - Flow control over runs
 14:                    0 - Invalid packets
 15:                    0 - Errors
 16:                    0 - Short sends
 17:               55,230 - Client keep-alives
 18:                  433 - Sender flow control limits applied
 19:                    0 - Unblocked Publications
 20:                    0 - Unblocked Control Commands
 21:                    0 - Possible TTL Asymmetry

1306:               65,792 - pub-lmt: 896 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
 1307:               33,024 - snd-pos: 896 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
 1480:               33,024 - sub-pos: 76 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1481:               33,024 - sub-pos: 201 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1482:               33,024 - sub-pos: 265 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1483:               33,024 - sub-pos: 302 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1484:               33,024 - sub-pos: 379 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1485:               33,024 - sub-pos: 491 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1486:               33,024 - sub-pos: 711 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1487:               33,024 - sub-pos: 770 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1488:                  576 - sub-pos: 1136 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @576
 1489:               33,024 - rcv-hwm: 1182 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
 1490:               33,024 - rcv-pos: 1182 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
 1513:               33,024 - sub-pos: 1207 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @24,160
 1578:               33,024 - sub-pos: 1261 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @24,256
 1906:               33,024 - sub-pos: 1492 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @33,024
 2090:               33,024 - sub-pos: 1639 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @33,024
 2217:               33,024 - sub-pos: 1745 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @33,024
 2332:               33,024 - sub-pos: 1848 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @33,024
 2360:               33,024 - sub-pos: 1875 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @33,024
 2391:               33,024 - sub-pos: 1913 926135193 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @33,024
```

The log buffer data ([aeron-buffer.zip](https://github.com/real-logic/Aeron/files/331136/aeron-buffer.zip)) is also available.

Is there any other data that would be useful to capture?

**Comment by tmontgomery** [comment]
Thanks, Mike.

Looks like the everything is normal, actually. The joining position is 576. Which is where registrationId 1136 is located. Others have this same joining position and moved on just fine. The rcv-pos is equal to rcv-hwm is equal to snd-pos. So, there is no loss detected from 576 onward. Which means that delivery should have continued just fine.

It would be good to know if that subscription has continued to poll.

I'll take a look at extending LogInspector to handle logbuffers that start with zero as this one will as the lowest joining position is 576.

Something interesting is that the joining position of several subscriptions is the current snd-pos, etc. Nothing wrong with that, really. Just interesting.

As the lowest join position is 576, this means that there is/was a separate media driver for this stream or that a previous run on this media driver used this stream and moved it from 0 to 576.

**Comment by tmontgomery** [comment]
I've added an option to LogInspector to scan over zeroes to take a look at the Image file.

There is a frame at offset 576 in Term index 0. Which is the active index. So, there is data at the position.

So, given that there is data at the position, the only ways I can see poll not delivering are:

1. poll is not called. Unlikely given what you have monitored.
1. this image isn't the one being polled somehow. Maybe weirdness of mapping?

One thing that might be useful would be the sequence of commands that show Available/Unavailable images. To help with the later above.

Something I noticed is that some messages at the Publication are 0 length. While not an issue, I wanted to mention it in case it isn't intentional.

**Comment by tmontgomery** [comment]
It would be good to know the sequence of pub/sub adds/removes here as well as the available/unavailable images. It looks from the registrationIds and join positions that there is a good bit of add/remove of subs and pubs around this streamId going on.

**Comment by tmontgomery** [comment]
@mikeb01 are you checking the return value from `Subscription::poll`?

**Comment by mikeb01** [comment]
Thanks @tmontgomery.  The subscription should still be polled.  I forgot to capture it on the latest run, but our code hasn't changed since the previous so I would expect it to be.

I'm assuming that the information around pub/sub adds/removes as well as the available/unavailable images is available via logging.  I'll have a look at getting that switched on and do another run.

**Comment by mikeb01** [comment]
I'll look to capture the following:

```
CMD_IN_ADD_PUBLICATION
CMD_IN_REMOVE_PUBLICATION
CMD_IN_ADD_SUBSCRIPTION
CMD_IN_REMOVE_SUBSCRIPTION
CMD_OUT_PUBLICATION_READY
CMD_OUT_AVAILABLE_IMAGE
CMD_OUT_ON_UNAVAILABLE_IMAGE
```

**Comment by tmontgomery** [comment]
Thanks, @mikeb01 Those should work fine. Remember that the logging now uses the aeron-agent if you haven't used it in a while.

**Comment by mikeb01** [comment]
Just look at that now.  Is there a reason why you configure the agent using a system property?  The typical approach for configuring the java would be:
```
-javaagent:agent.jar=aeron.event.log=admin,FRAME_IN
```

**Comment by tmontgomery** [comment]
Simply didn't get around to doing it.

**Comment by mjpt777** [comment]
I have a suspicion that something might be corrupting the log. I've pushed a new feature that allows images to be mapped read-only which should catch this. It is now the default and can be changed from the client context.

**Comment by tmontgomery** [comment]
@mikeb01 is it possible that an exception is being seen on the fragment handler for that stream?

Also, are the 0 length messages intentional?

**Comment by mikeb01** [comment]
I'll do some more digging on Friday.  I think the zero length messages are intentional, but I will need to double check.  I'll also run through the code to make sure there are no silent exceptions occurring.

**Comment by tmontgomery** [comment]
The only other thing that is suspicious is that the joining positions are non-0. This can only really happen if there is another Media Driver and subscription that is listening and/or a different Flow Control strategy or channel endpoint.

**Comment by mikeb01** [comment]
Are the joining positions, the values after the `@` in AeronStat?

**Comment by mjpt777** [comment]
Yes

**Comment by mikeb01** [comment]
I don't think that is too big a deal.  With out system it is possible that a subscription could occur after publication has been created and messages have already been sent.  In that case our late join capability should kick in to provide redeliver messages where enabled and in other cases the message loss is inconsequential.

I'm assuming that for a multicast channel, a subscription can join a stream at any point and receive messages from that point onward.  Messages delivered prior to joining won't be seen by the new subscription, but for messages sent after the new subscription will be included in the set of Subscriptions that can create back-pressure.

**Comment by tmontgomery** [comment]
@mikeb01 correct. Joining in is fine. And gives the semantics you mention.

But to get into that state, another media driver must be around... or flow control strategy, etc.

**Comment by mikeb01** [comment]
I'll have to compare to what I see locally.  I am running ~60 instances of the application each with a separate media driver on separate virtual machines.  Each is using a dummy interface to keep all of the multicast traffic local to that instance.  However they will be sharing multicast addresses so I'll make sure that multicast traffic is not leaking across somehow.

**Comment by tmontgomery** [comment]
I don't believe this should be a problem. Just something I have noticed. While trying to simulate what you are seeing, I have been trying to make things happen, etc.

**Comment by tmontgomery** [comment]
@mikeb01 is there any exceptions in the Aeron client error handler for any of the clients? or any evidence of FD or VM exhaustion in the applications?

@mjpt777 and I are talking about possibly the mappings being 0 due to VM or FD exhaustion somehow.

**Comment by mikeb01** [comment]
Checking my acronyms, FD == file descriptor and VM == virtual memory?  I'll double check the file descriptor ulimits, but they should be okay.  As for memory, we have 16GB allocated to /dev/shm and during a run have used <1GB of that.  I'll grab some more data when I do some more runs through tomorrow.

**Comment by tmontgomery** [comment]
correct on acronyms, yes.

**Comment by mikeb01** [comment]
Another run through today.  File descriptors and memory looks okay

```
[barkerm@lddev-build-sas018 ~]$ sysctl fs.file-nr
fs.file-nr = 22560	0	3265313
[barkerm@lddev-build-sas018 ~]$ pgrep -f name=aeron-media-driver
22694
[barkerm@lddev-build-sas018 ~]$ lsof -a -p 22694 |  wc -l
5
[barkerm@lddev-build-sas018 ~]$ cat /proc/meminfo 
MemTotal:       32878504 kB
MemFree:        16942308 kB
Buffers:          329116 kB
Cached:          3647380 kB
...
```
For this run it was stream `-215825341` that failed.
AeronStat:
```
01:35:57 - Aeron Stat
=========================
  0:          101,125,472 - Bytes sent
  1:          103,091,488 - Bytes received
  2:                    0 - Failed offers to ReceiverProxy
  3:                    0 - Failed offers to SenderProxy
  4:                    0 - Failed offers to DriverConductorProxy
  5:               14,199 - NAKs sent
  6:               14,199 - NAKs received
  7:            1,475,304 - Status Messages sent
  8:            1,475,030 - Status Messages received
  9:            2,870,423 - Heartbeats sent
 10:            2,781,648 - Heartbeats received
 11:                4,989 - Retransmits sent
 12:                  831 - Flow control under runs
 13:                    0 - Flow control over runs
 14:                    0 - Invalid packets
 15:                    0 - Errors
 16:                    0 - Short sends
 17:               58,353 - Client keep-alives
 18:                  433 - Sender flow control limits applied
 19:                    0 - Unblocked Publications
 20:                    0 - Unblocked Control Commands
 21:                    0 - Possible TTL Asymmetry

78:               64,928 - pub-lmt: 48 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
79:               32,160 - snd-pos: 48 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
1285:               32,160 - rcv-hwm: 1104 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
1286:               32,160 - rcv-pos: 1104 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
1289:               32,160 - sub-pos: 1017 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1327:                  288 - sub-pos: 1127 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @288
1873:               32,160 - sub-pos: 1374 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @6,848
2115:               32,160 - sub-pos: 1583 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @32,160
2211:               32,160 - sub-pos: 1658 1039395103 -215825341 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @32,160
```

I've included the publication and image files, along with the event log in 
[aeron-data.zip](https://github.com/real-logic/Aeron/files/342749/aeron-data.zip).

Interestingly the image does go unavailable during the run, then becomes available again.
```
[3225040.590694] CMD_OUT_AVAILABLE_IMAGE [150/150]: 1039395103:-215825341 [0:1470:1017] "172.29.1.1:32933" [1018]
[3225050.909230] CMD_OUT_ON_UNAVAILABLE_IMAGE [73/73]: aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 -215825341 [1018]
[3225051.761203] CMD_OUT_AVAILABLE_IMAGE [150/150]: 1039395103:-215825341 [0:1289:1017] "172.29.1.1:32933" [1104]
[3225054.078744] CMD_OUT_AVAILABLE_IMAGE [150/150]: 1039395103:-215825341 [0:1329:1127] "172.29.1.1:32933" [1018]
[3225054.078805] CMD_OUT_AVAILABLE_IMAGE [150/150]: 1039395103:-215825341 [0:1327:1127] "172.29.1.1:32933" [1104]
[3225079.567818] CMD_OUT_AVAILABLE_IMAGE [150/150]: 1039395103:-215825341 [0:1873:1374] "172.29.1.1:32933" [1104]
[3225114.814036] CMD_OUT_AVAILABLE_IMAGE [150/150]: 1039395103:-215825341 [0:2115:1583] "172.29.1.1:32933" [1104]
[3225123.918616] CMD_OUT_AVAILABLE_IMAGE [150/150]: 1039395103:-215825341 [0:2211:1658] "172.29.1.1:32933" [1104]
```

I've also added a bunch of additional exception handling on the subscriber side, with logging.  The logs are quiet, so it doesn't seem to be an exception in the handler breaking it.

I also tracked down the zero length messages.  The run that I've included the data from still has those messages, but I've had subsequent runs without the zero length messages and I still see the same behaviour.

**Comment by mikeb01** [comment]
Retesting with 0.9.10-SNAPSHOT, still seeing the same issue:

```
15:44:09 - Aeron Stat
=========================
  0:           95,839,872 - Bytes sent
  1:           97,988,320 - Bytes received
  2:                    0 - Failed offers to ReceiverProxy
  3:                    0 - Failed offers to SenderProxy
  4:                    0 - Failed offers to DriverConductorProxy
  5:               14,815 - NAKs sent
  6:               14,815 - NAKs received
  7:            1,434,072 - Status Messages sent
  8:            1,434,250 - Status Messages received
  9:            2,785,669 - Heartbeats sent
 10:            2,695,995 - Heartbeats received
 11:                5,692 - Retransmits sent
 12:                  239 - Flow control under runs
 13:                    0 - Flow control over runs
 14:                    0 - Invalid packets
 15:                    0 - Errors
 16:                    0 - Short sends
 17:               57,004 - Client keep-alives
 18:                  423 - Sender flow control limits applied
 19:                    0 - Unblocked Publications
 20:                    0 - Unblocked Control Commands
 21:                    0 - Possible TTL Asymmetry
 22:                    1 - send-channel: aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1

721:               64,800 - pub-lmt: 880 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
722:               32,032 - snd-pos: 880 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
858:               32,032 - sub-pos: 1153 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @64
1323:               32,032 - sub-pos: 73 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1324:               32,032 - sub-pos: 171 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1325:               32,032 - sub-pos: 276 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1326:               32,032 - sub-pos: 300 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1327:               32,032 - sub-pos: 382 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1328:               32,032 - sub-pos: 542 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1329:               32,032 - sub-pos: 848 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1330:               32,032 - sub-pos: 855 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1331:               32,032 - rcv-hwm: 1009 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
1332:               32,032 - rcv-pos: 1009 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 
1459:                    0 - sub-pos: 1099 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @0
1553:               32,032 - sub-pos: 1213 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @8,320
1794:               32,032 - sub-pos: 1358 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @32,032
2073:               32,032 - sub-pos: 1575 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @32,032
2211:               32,032 - sub-pos: 1685 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @32,032
2320:               32,032 - sub-pos: 1786 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @32,032
2350:               32,032 - sub-pos: 1812 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @32,032
2375:               32,032 - sub-pos: 1848 -747111497 -1762877036 aeron:udp?endpoint=239.193.0.1:10000|interface=172.29.1.1 @32,032
```

**Comment by mikeb01** [comment]
Increased imageLivenessTimeout to 30 seconds and the system appears to work.

**Comment by tmontgomery** [comment]
@mikeb01 I assume the image unavailability also stops with upping the liveness?

**Comment by tmontgomery** [comment]
@mikeb01 back a while ago, you identified that an unavailable image was causing data to stop. At the time we tracked that down to SETUP not being used when data was not sent causing things to stop. I wonder, though, if somehow the unavailable image is not being handled appropriately.

You are calling `Subscription.poll` and not `Image.poll` directly, correct?

As images go unavailable and come back, they are new images. Polling on old ones will be fruitless. Subscription manages this. Or should. Which I have looked at a number of times in the last few weeks. Nothing I see should be perpetually out of whack.

**Comment by mikeb01** [comment]
@tmontgomery I unfortunately didn't get the logs on the run that managed to start, so I'm not sure if the unavailable image events went away or not.  I'll do some more runs and capture that information.

We are using `Subscription.poll`.  Let me know if there is any additional logging I can include to try and track down what it going on.  It is pretty easy for me to repeat the failure.

**Comment by mjpt777** [comment]
@mikeb01 I tracked CPU usage to a busy spinning setup thread when configuring a large number of streams. It can take a number of milliseconds to setup each new publication or subscription in creating log buffers and sockets. The client thread now sleeps awaiting the operation to finish and thus frees the CPU resource to carry out that task when in a contended environment such as CI.

**Comment by mikeb01** [comment]
I'll give a run next week.

**Comment by mjpt777** [comment]
@mikeb01 Also given you are using the shared threading mode in the driver for CI then the move to incremental cleaning should make things more responsive.

**Comment by tmontgomery** [comment]
Should also add a means to have the BroadcastReceiver on the client track the number of missed messages and expose it for clients to be able to see. For debugging, that is.

**Comment by mjpt777** [comment]
If the client cannot keep up with the broadcast buffer it gets an `IllegalStateException`. I assume this has not been observed in the client?

**Comment by mjpt777** [comment]
@mikeb01 The exception will propagate up the call stack if it occurs during operation to add or release a subscription/publication, or to the error handler if it occurs during the duty cycle of the client conductor.

**Comment by mjpt777** [comment]
The fact that a change in image liveness timeout makes me think it is worth investigating the state machine for image cleanup for when many subscriptions exist.

**Comment by tmontgomery** [comment]
Ah. The CopyBroadcastReceiver has the check. Was just looking at the BroadcastReceiver. Seems odd to place that behavior there to me at the moment. But that is probably just initial reaction since I wrote the C++ version a while ago (and it has the check also).

@mikeb01 do you have logs and logbuffers for the last run you had that demonstrated the issue? (i.e. not the clean run, but the bad one with the stats above). If not, that is OK. I just want more data to stare at.

Let me work through some logic...

The state machine has almost no dependency to subscriptions aside from event purposes. And, the logs look normal from an image management perspective. Unless I am missing something. And, the unavailable image event has nothing to do with the subscription position. It simply informs the client that the image is not available. Then the client matches the image correlation Id per Subscription. Thus, only a single event per image is needed. If state management of the image were a problem in the driver, the log would show something odd. If state management of the image were a problem in the client, then it would be fairly obvious since the logic is quite simple and is totally driven by events.

The thread starvation causing an image to go unavailable is a temporary thing and the system should eventually resume. It could be pathological, though. 5 seconds of no DATA or SETUP is a long time. And loss indications point to a lot of congestion and/or starvation of the Receiver thread. But the stats indicate there is data received and delivered to all but one subscriber. Which means the data does make it into the logbuffer. And the logbuffer files indicate that also.

**Comment by tmontgomery** [comment]
Hmmm. I think I might see something. Another event log might help to confirm.

addImage looks for sessionId. But I see on an addSubscription that the same sessionId is being added twice. Which is fine for situations where there is an image going away and being re-instated. Which is this case. But it is on a new subscription.

This might be a timing thing related to adding a subscription while one is in linger. Somewhat like you mention, @mjpt777 , but actually, not. :)

**Comment by tmontgomery** [comment]
The anomaly in the log indicates some potential issues. It may not be _the_ issue, though. But it could explain the behavior. Would rather have some more event logs and stats of bad runs to solidify.

What I am thinking is that relying on sessionId for addImage is the wrong thing to do on the client. It needs to be correlationId instead.

That is one thing, but there is another.

Another is that sending unavailable image on going into linger is fine, but then it is imperative to not send available image after that when doing an addSubscription. This must be coordinated in the driver between Receiver and Conductor threads correctly. Or waiting for unavailable image to just before cleanup.

**Comment by mikeb01** [comment]
I don't have any to hand, but I can produce some later this week.

**Comment by tmontgomery** [comment]
I've done 6f2d1f14d02ce0fce825ba6eb1b560f9eb0b3d92 to try and avoid issuing an availableImage when doing an addSubscription on an image that is not `ACTIVE`. This should help avoid the case seen in the one event log that could explain things.

I don't know if this also addresses anything else seen, though.

**Comment by tmontgomery** [comment]
Convinced myself to go ahead and make the addImage filtering be correlationId based. bbaf75dba0d6f42435d1ca9bfc9628559e85c9ed

**Comment by mikeb01** [comment]
Which version do you want me to gather data from?  Latest release or latest snapshot?

**Comment by tmontgomery** [comment]
Just trying out latest snapshot should say whether things are working better or not. Gathering any logbuffers and event logs and stats from it should help.

**Comment by mikeb01** [comment]
I think this is now fixed.  3 runs and the problem is yet to recur.  Not sure if it is due to less aggressive spinning (b8d83be5baeb4eb731a6cd632d5039204b173fb3) or using correlationIds in addImage (bbaf75dba0d6f42435d1ca9bfc9628559e85c9ed).

**Comment by mjpt777** [comment]
Thanks @mikeb01 all the feedback has been really helpful.

**Comment by tmontgomery** [comment]
Thanks, @mikeb01 !
